analysis/create_project_actions.R

```diff
@@ -0,0 +1,148 @@
+    str_replace_all("\\#\\#\\'\\\n", "\n")
+}
+
+# Create function to generate study population ---------------------------------
+
+generate_study_population <- function(cohort){
+  splice(
+    comment(glue("Generate study population - {cohort}")),
+    action(
+      name = glue("generate_study_population_{cohort}"),
+      run = glue("ehrql:v1 generate-dataset analysis/dataset_definition_{cohort}.py --output output/input_{cohort}.csv.gz"),
+      needs = list("generate_dataset_index_dates"),
+      highly_sensitive = list(
+        cohort = glue("output/input_{cohort}.csv.gz")
+      )
+    )
+  )
+}
+
+
+# Define and combine all actions into a list of actions ------------------------
+
+actions_list <- splice(
+  
+  ## Post YAML disclaimer ------------------------------------------------------
+  
+  comment("# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #",
+          "DO NOT EDIT project.yaml DIRECTLY",
+          "This file is created by create_project_actions.R",
+          "Edit and run create_project_actions.R to update the project.yaml",
+          "# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #"
+  ),
+  
+  ## Generate vaccination eligibility information ------------------------------
+  comment("Generate vaccination eligibility information"),
+  
+  action(
+    name = glue("vax_eligibility_inputs"),
+    run = "r:latest analysis/metadates.R",
+    highly_sensitive = list(
+      study_dates_json = glue("output/study_dates.json")
+    )
+  ),
+  
+  ## Generate index dates for all study cohorts ------------------------------------------
+  comment("Generate dates for all cohorts"),
+  
+  action(
+    name = "generate_dataset_index_dates",
+    run = "ehrql:v1 generate-dataset analysis/dataset_definition_dates.py --output output/index_dates.csv.gz",
+    needs = list("vax_eligibility_inputs"),
+    highly_sensitive = list(
+      dataset = glue("output/index_dates.csv.gz")
+    )
+  ),
+  
+  ## Generate study population -------------------------------------------------
+  
+  splice(
+    unlist(lapply(cohorts, 
+                  function(x) generate_study_population(cohort = x)), 
+           recursive = FALSE
+    )
+  )
+)
+
+# Combine actions into project list --------------------------------------------
+
+project_list <- splice(
+  defaults_list,
+  list(actions = actions_list)
+)
+
+# Convert list to yaml, reformat, and output a .yaml file ----------------------
+
+as.yaml(project_list, indent=2) %>%
+  # convert comment actions to comments
+  convert_comment_actions() %>%
+  # add one blank line before level 1 and level 2 keys
+  str_replace_all("\\\n(\\w)", "\n\n\\1") %>%
+  str_replace_all("\\\n\\s\\s(\\w)", "\n\n  \\1") %>%
+  writeLines("project.yaml")
+
+# Return number of actions -----------------------------------------------------
+
+print(paste0("YAML created with ",length(actions_list)," actions."))
```

Comment:
It looks like we made a mistake with this previously as it is saying 15 actions when you have 5 actions. The following code should fix it:

```
# Return number of actions -----------------------------------------------------

count_run_elements <- function(x) {

  if (!is.list(x)) {
    return(0)
  }

  # Check if any names of this list are "run"
  current_count <- sum(names(x) == "run", na.rm = TRUE)
  
  # Recursively check all elements in the list
  return(current_count + sum(sapply(x, count_run_elements)))

}

print(paste0("YAML created with ",count_run_elements(actions_list)," actions."))
```